apphub/instance_detection/yolov5/yolov5_tf.py

```diff
@@ -0,0 +1,579 @@
+        self.loss_conf = tf.losses.BinaryCrossentropy(reduction="none")
+        self.loss_cls = tf.losses.BinaryCrossentropy(reduction="none")
+        self.img_size = img_size
+
+    def forward(self, data, state):
+        pred, true = data
+        true_box, true_obj, true_class = tf.split(true, (4, 1, -1), axis=-1)
+        pred_box, pred_obj, pred_class = tf.split(pred, (4, 1, -1), axis=-1)
+        num_classes = pred_class.shape[-1]
+        true_class = tf.squeeze(tf.one_hot(tf.cast(true_class, tf.int32), depth=num_classes, axis=-1), -2)
+        box_scale = 2 - 1.0 * true_box[..., 2] * true_box[..., 3] / (self.img_size**2)
+        obj_mask = tf.squeeze(true_obj, -1)
+        conf_focal = tf.squeeze(tf.math.pow(true_obj - pred_obj, 2), -1)
+        iou = self.bbox_iou(pred_box, true_box, giou=True)
+        iou_loss = (1 - iou) * obj_mask * box_scale
+        conf_loss = conf_focal * self.loss_conf(true_obj, pred_obj)
+        class_loss = obj_mask * self.loss_cls(true_class, pred_class)
+        iou_loss = tf.reduce_mean(tf.reduce_sum(iou_loss, axis=[1, 2, 3]))
+        conf_loss = tf.reduce_mean(tf.reduce_sum(conf_loss, axis=[1, 2, 3]))
+        class_loss = tf.reduce_mean(tf.reduce_sum(class_loss, axis=[1, 2, 3])) * num_classes
+        return iou_loss, conf_loss, class_loss
+
+    @staticmethod
+    def bbox_iou(bbox1, bbox2, giou=False, diou=False, ciou=False, epsilon=1e-7):
+        b1x1, b1x2, b1y1, b1y2 = bbox1[..., 0], bbox1[..., 0] + bbox1[..., 2], bbox1[..., 1], bbox1[..., 1] + bbox1[..., 3]
+        b2x1, b2x2, b2y1, b2y2  = bbox2[..., 0], bbox2[..., 0] + bbox2[..., 2], bbox2[..., 1], bbox2[..., 1] + bbox2[..., 3]
+        # intersection area
+        inter = tf.maximum(tf.minimum(b1x2, b2x2) - tf.maximum(b1x1, b2x1), 0) * tf.maximum(
+            tf.minimum(b1y2, b2y2) - tf.maximum(b1y1, b2y1), 0)
+        # union area
+        w1, h1 = b1x2 - b1x1 + epsilon, b1y2 - b1y1 + epsilon
+        w2, h2 = b2x2 - b2x1 + epsilon, b2y2 - b2y1 + epsilon
+        union = w1 * h1 + w2 * h2 - inter + epsilon
+        # iou
+        iou = inter / union
+        if giou or diou or ciou:
+            # enclosing box
+            cw = tf.maximum(b1x2, b2x2) - tf.minimum(b1x1, b2x1)
+            ch = tf.maximum(b1y2, b2y2) - tf.minimum(b1y1, b2y1)
+            if giou:
+                enclose_area = cw * ch + epsilon
+                return iou - (enclose_area - union) / enclose_area
+            if diou or ciou:
+                c2 = cw**2 + ch**2 + epsilon
+                rho2 = ((b2x1 + b2x2) - (b1x1 + b1x2))**2 / 4 + ((b2y1 + b2y2) - (b1y1 + b1y2))**2 / 4
+                if diou:
+                    return iou - rho2 / c2
+                elif ciou:
+                    v = (4 / math.pi**2) * tf.pow(tf.atan(w2 / h2) - tf.atan(w1 / h1), 2)
+                    alpha = v / (1 - iou + v)
+                    return iou - (rho2 / c2 + v * alpha)
+        return tf.clip_by_value(iou, 0, 1)
+
+
+class Rescale(TensorOp):
```

Comment:
Could just use the Normalize op w/ mean=0, std=1

Reply:
but it'd be less intuitive for new user trying to learn what is going on. Since they need to know Normalize divides value by 255 under the hood.